crates/proof-of-sql/src/sql/postprocessing/group_by_postprocessing.rs

```diff
@@ -0,0 +1,322 @@
+};
+use serde::{Deserialize, Serialize};
+
+/// A tuple of an aggregation operator, an expression, and an identifier
+pub type AggregationTuple = (AggregationOperator, Expression, Identifier);
+
+/// A group by expression
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct GroupByPostprocessing {
+    /// A list of `AliasedResultExpr` that exclusively use identifiers in the group by clause or results of aggregation expressions
+    remainder_exprs: Vec<AliasedResultExpr>,
+
+    /// A list of identifiers in the group by clause
+    group_by_identifiers: Vec<Identifier>,
+
+    /// A list of aggregation expressions
+    agg_tuples: Vec<AggregationTuple>,
+}
+
+/// Check whether multiple layers of aggregation exist within the same GROUP BY clause
+/// since this is not allowed in SQL
+///
+/// If the context is within an aggregation function, then any aggregation function is considered nested.
+/// Otherwise we need two layers of aggregation functions to be nested.
+pub(super) fn contains_nested_aggregation(expr: &Expression, is_agg: bool) -> bool {
+    match expr {
+        Expression::Column(_) | Expression::Literal(_) | Expression::Wildcard => false,
+        Expression::Aggregation { expr, .. } => {
+            if is_agg {
+                true
+            } else {
+                contains_nested_aggregation(expr, true)
+            }
+        }
+        Expression::Binary { left, right, .. } => {
+            contains_nested_aggregation(left, is_agg) || contains_nested_aggregation(right, is_agg)
+        }
+        Expression::Unary { expr, .. } => contains_nested_aggregation(expr, is_agg),
+    }
+}
+
+/// Get identifiers NOT in aggregate functions
+pub(super) fn get_free_identifiers_from_expr(expr: &Expression) -> IndexSet<Identifier> {
+    match expr {
+        Expression::Column(identifier) => IndexSet::from([*identifier]),
+        Expression::Literal(_) | Expression::Aggregation { .. } | Expression::Wildcard => {
+            IndexSet::new()
+        }
+        Expression::Binary { left, right, .. } => {
+            let left_identifiers = get_free_identifiers_from_expr(left);
+            let right_identifiers = get_free_identifiers_from_expr(right);
+            left_identifiers
+                .union(&right_identifiers)
+                .cloned()
+                .collect::<IndexSet<_>>()
+        }
+        Expression::Unary { expr, .. } => get_free_identifiers_from_expr(expr),
+    }
+}
+
+/// Get aggregate expressions from an expression as well as the remainder
+///
+/// The idea here is to recursively traverse the expression tree and collect all the aggregation expressions
+/// and then label them as new columns post-aggregation and replace them with these new columns so that
+/// the post-aggregation expression tree doesn't contain any aggregation expressions and can be simply evaluated.
+pub(super) fn get_aggregate_and_remainder_expressions(
+    expr: &Expression,
+    agg_count: &mut usize,
```

Comment:
@JayWhite2357 Well the reason why we need this is that we need to cover cases such as `SUM(a + 1) + MAX(b - c) + d as res`. The idea here is that we replace each aggregate function with a new column on the post-aggregation table, that is, this will become `__col_agg_0 + __col_agg_1 + d as res` with `__col_agg_0` and `__col_agg_1` being aggregation outcome of `SUM(a + 1)` and `MAX(b - c)` respectively. So at the very least we need `(Vec<(AggregationOperator, Expression)>, Expression)`. I use numbers to label these new columns because stateless alternatives such as using stringified versions of intermediate ASTs will lead to issues with the `Identifier` length limit.